runtime/service.go

```diff
@@ -1183,42 +1184,58 @@ func (s *service) shutdown(
 	defer cancel()
 	go func() {
-		// Once the shutdown procedure is done, the shim needs to shutdown too.
-		// This also ensures that if the VMM is still alive, it will receive a
-		// SIGKILL via exec.CommandContext
-		<-shutdownCtx.Done()
+		// Once the firecracker process is complete, the shim needs to shutdown too.
+		err := s.machine.Wait(context.Background())
+		if err != nil {
+			s.logger.WithError(err).Error("the VMM didn't finish successfully")
+		}
 		s.shimCancel()
 	}()
 
 	s.logger.Info("stopping the VM")
 
-	// Try to tell agent to exit, causing the VM to begin a reboot. If that
-	// fails, try to forcibly stop the VMM. If that too fails, just cancel
-	// the shutdownCtx to fast-path to the VMM getting SIGKILL.
-	_, shutdownErr := s.agentClient.Shutdown(shutdownCtx, req)
-	if shutdownErr != nil {
-		s.logger.WithError(shutdownErr).Error("failed to shutdown VM agent")
-		stopVMMErr := s.machine.StopVMM()
-		if stopVMMErr != nil {
-			s.logger.WithError(stopVMMErr).Error("failed to forcibly stop VMM")
-			cancel()
+	// 1) Tell agent to exit, causing the VM to begin a reboot.
+	_, err := s.agentClient.Shutdown(shutdownCtx, req)
+	if err != nil {
+		s.logger.WithError(err).Error("failed to shutdown VM agent")
+	} else {
+		// If the agent can take the request, wait a moment to gracefully shutdown.
+		err = s.machine.Wait(shutdownCtx)
+		if err != nil {
+			s.logger.WithError(err).Error("failed to wait the VMM")
+		} else {
+			return nil // Success!
 		}
 	}
 
-	// wait for the shimCtx to be done, which means the VM has exited and we're ready
-	// to shutdown
-	<-s.shimCtx.Done()
-	if shutdownCtx.Err() == context.DeadlineExceeded {
-		return status.Error(codes.DeadlineExceeded,
-			"timed out waiting for VM shutdown, VMM was sent SIGKILL")
+	// If Wait() returns before the deadline, the VMM might be killed by someone else
+	if shutdownCtx.Err() != context.DeadlineExceeded {
+		if s.vmExitErr != nil {
+			return status.Error(codes.Internal, fmt.Sprintf("VMM exit errors: %v", s.vmExitErr))
+		}
 	}
 
-	if s.vmExitErr != nil {
-		return status.Error(codes.Internal,
-			fmt.Sprintf("VMM exit errors: %v", s.vmExitErr))
+	// 2) If that fails, stop the VMM by sending SIGTERM
+	waitCtx, waitCancel := context.WithTimeout(requestCtx, defaultSigtermTimeout)
+	defer waitCancel()
+
+	err = s.machine.StopVMM()
+	if err != nil {
+		s.logger.WithError(err).Error("failed to stop the VM")
 	}
 
-	return nil
+	err = s.machine.Wait(waitCtx)
+	if waitCtx.Err() != context.DeadlineExceeded {
+		return status.Error(codes.DeadlineExceeded, fmt.Sprintf("the VMM was killed by SIGTERM: %v", err))
+	}
+
+	// 3) If that fails, cancelling the context to send SIGKILL
+	s.shimCancel()
+	err = s.machine.Wait(context.Background())
+	if err != nil {
+		s.logger.WithError(err).Error("failed to wait the VMM after cancelling the context")
+	}
+	return status.Error(codes.DeadlineExceeded, fmt.Sprintf("the VMM was killed by SIGKILL: %v", err))
```

Comment:
Yeah, I agree that DeadlineExceeded is a bit odd, since the specified action (stopping the VM) is executed anyway. What I wanted to tell were

- The shutdown process is exceeding the customer-specified timeout.
- Because of that, the firecracker process is killed by an undesirable way.

Reply:
Looking at the other codes the only two that seemed like maybe-appropriate were Unknown and Internal. Do either of those seem appropriate here?

Reply:
I'm fine with `Internal`. Let me update the implementation.

Reply:
Done.